numpy/_core/src/umath/umathmodule.c

```diff
@@ -40,25 +40,133 @@
+        PyArrayMethodObject *self,
+        PyArray_DTypeMeta *const *NPY_UNUSED(dtypes),
+        PyArray_Descr *const *NPY_UNUSED(given_descrs),
+        PyArray_Descr **loop_descrs,
+        npy_intp *NPY_UNUSED(view_offset))
+{
+    int i, nop = self->nin + self->nout;
+
+    loop_descrs[0] = PyArray_DescrFromType(NPY_OBJECT);
+    if (loop_descrs[0] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+
+    for (i = 1; i < nop; ++i) {
+        Py_INCREF(loop_descrs[0]);
+        loop_descrs[i] = loop_descrs[0];
+    }
+
+    return NPY_NO_CASTING;
+}
+
+static int
+frompyfunc_promoter(PyObject *ufunc,
+        PyArray_DTypeMeta *const NPY_UNUSED(op_dtypes[]),
+        PyArray_DTypeMeta *const NPY_UNUSED(signature[]),
+        PyArray_DTypeMeta *new_op_dtypes[]);
 
 static int
-object_ufunc_type_resolver(PyUFuncObject *ufunc,
-                                NPY_CASTING casting,
-                                PyArrayObject **operands,
-                                PyObject *type_tup,
-                                PyArray_Descr **out_dtypes)
+frompyfunc_promoter(PyObject *ufunc,
+        PyArray_DTypeMeta *const NPY_UNUSED(op_dtypes[]),
+        PyArray_DTypeMeta *const NPY_UNUSED(signature[]),
+        PyArray_DTypeMeta *new_op_dtypes[])
 {
-    int i, nop = ufunc->nin + ufunc->nout;
+    int i, nop = ((PyUFuncObject *)ufunc)->nin + ((PyUFuncObject *)ufunc)->nout;
 
-    out_dtypes[0] = PyArray_DescrFromType(NPY_OBJECT);
-    if (out_dtypes[0] == NULL) {
+    new_op_dtypes[0] = PyArray_DTypeFromTypeNum(NPY_OBJECT);
+    if (new_op_dtypes[0] == NULL) {
         return -1;
     }
 
     for (i = 1; i < nop; ++i) {
-        Py_INCREF(out_dtypes[0]);
-        out_dtypes[i] = out_dtypes[0];
+        Py_INCREF(new_op_dtypes[0]);
+        new_op_dtypes[i] = new_op_dtypes[0];
+    }
+
+    return 0;
+}
+
+static int
+pyfunc_loop(PyArrayMethod_Context *context,
+                char *const data[], npy_intp const dimensions[],
+                npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    npy_intp N = dimensions[0];
+    int nin = context->method->nin;
+    int nout = context->method->nout;
+    int ntot = nin + nout;
+    PyObject *pyfunc = context->method->static_data;
+    char *ptrs[NPY_MAXARGS];
+    PyObject *arglist, *result;
+    PyObject *in, **op;
+    npy_intp i;
+
+    for(i = 0; i < ntot; i++) {
+        ptrs[i] = data[i];
+    }
+    while(N--) {
+        arglist = PyTuple_New(nin);
+        if (arglist == NULL) {
+            return -1;
+        }
+        for(i = 0; i < nin; i++) {
+            in = *((PyObject **)ptrs[i]);
+            /* We allow NULL, but try to guarantee non-NULL to downstream */
+            assert(in != NULL);
+            if (in == NULL) {
+                in = Py_None;
+            }
+            PyTuple_SET_ITEM(arglist, i, in);
+            Py_INCREF(in);
+        }
+        result = PyObject_CallObject(pyfunc, arglist);
+        Py_DECREF(arglist);
```

Comment:
Arg, sorry.  You will still need to make a copy of all arguments because of the `None`/`NULL` handling.  So best we can do is a `args[NPY_MAXARGS]` assignment.

So if you prefer to ignore it, please feel free.